src/ir_passes/mono/utils.rs

```diff
@@ -1,4 +1,125 @@
-use crate::utils::{self, Idx};
+use crate::{
+    ast,
+    ir::{self, Ctx, IndexStore, InterfaceSrc, Interned, MutCtx},
+    utils::{self, Idx},
+};
+
+pub struct UnderlyingComp<'a>(&'a ir::Component);
+
+impl<'a> UnderlyingComp<'a> {
+    pub fn new(comp: &'a ir::Component) -> Self {
+        Self(comp)
+    }
+    pub fn comp(&self) -> &ir::Component {
+        self.0
+    }
+    pub fn cmds(&self) -> &Vec<ir::Command> {
+        &self.0.cmds
+    }
+    pub fn is_ext(&self) -> bool {
+        self.0.is_ext
+    }
+    pub fn events(&self) -> &IndexStore<ir::Event> {
+        self.0.events()
+    }
+    pub fn exprs(&self) -> &Interned<ir::Expr> {
+        self.0.exprs()
+    }
+    pub fn params(&self) -> &IndexStore<ir::Param> {
+        self.0.params()
+    }
+    pub fn ports(&self) -> &IndexStore<ir::Port> {
+        self.0.ports()
+    }
+    pub fn src_info(&self) -> &Option<InterfaceSrc> {
+        &self.0.src_info
+    }
+    pub fn unannotated_ports(&self) -> &Vec<(ast::Id, u64)> {
+        &self.0.unannotated_ports
+    }
+}
+
+pub trait UnderlyingCtx<T> {
+    fn get(&self, k: Underlying<T>) -> &T;
+}
+
+impl<'a, T> UnderlyingCtx<T> for UnderlyingComp<'a>
+where
+    ir::Component: Ctx<T>,
+{
+    fn get(&self, k: Underlying<T>) -> &T {
+        self.0.get(k.idx())
+    }
+}
+
+pub struct BaseComp(ir::Component);
+
+impl BaseComp {
+    pub fn new(comp: ir::Component) -> Self {
+        Self(comp)
+    }
+
+    pub fn comp(&self) -> &ir::Component {
+        &self.0
+    }
+
+    pub fn set_unannotated_ports(&mut self, other: Vec<(ast::Id, u64)>) {
+        self.0.unannotated_ports = Box::new(other);
+    }
+    pub fn set_src_info(&mut self, other: Option<InterfaceSrc>) {
+        self.0.src_info = other;
+    }
+
+    pub fn extend_one(&mut self, other: Option<ir::Command>) {
+        self.0.cmds.extend(other);
+    }
+    pub fn extend_cmds(&mut self, other: Vec<ir::Command>) {
+        self.0.cmds.extend(other);
+    }
+    pub fn num(&mut self, n: u64) -> Base<ir::Expr> {
+        Base::new(self.0.add(ir::Expr::Concrete(n)))
+    }
+    pub fn bin(&mut self, expr: ir::Expr) -> Base<ir::Expr> {
+        Base::new(self.0.bin(expr))
+    }
+    pub fn func(&mut self, expr: ir::Expr) -> Base<ir::Expr> {
+        Base::new(self.0.func(expr))
+    }
+    pub fn resolve_prop(&mut self, prop: ir::Prop) -> Base<ir::Prop> {
+        Base::new(self.0.resolve_prop(prop))
+    }
+}
+
+pub trait BaseCtx<T> {
```

Comment:
Can we change the `Ctx` and `MutCtx` traits to take an `IdxLike` instead of operating on `Idx<T>`? This will allow us to implement `Ctx` for these new structs because `Base` and `Underlying` already implement `IdxLike`. The trick to doing this is changing the `Ctx` trait like this:
```rust
pub trait Ctx<T, K=Idx<T>> where K: IdxLike<T> {
    /// Add a new value to the context
    fn add(&mut self, val: T) -> K;
    /// Get the information associated with a value
    fn get(&self, idx: K -> &T;
}
```

This will allow all of the existing `Ctx` implementation to continue working because they will use the default value of `K`. However, for `BaseComp`, we can say:
```
impl Ctx<T, Base<T>> for BaseComp
  where ir::Component: Ctx<T> {
  fn get(&self, k: Base<T>) -> &T { self.0.get(k.get()) }
```

This is definitely getting to the galaxy brain meme levels of type magic so feel free to say no to this.

Reply:
Lol maybe, but I'll try it out!

Reply:
Also done in #288